### ISSUE TITLE
XBEL sync not working with owncloud (Origin header=null)

### ISSUE THREAD

**Issue by minj** [marcelklehr/floccus#153]
### Software versions
* Browser: Chrome 69.0.3497.42 beta (64-bit)
* Owncloud: 10.0.9 @ my.domain.tld/my_prefix
* Floccus: 3.0.2

### Sync settings
WebDAV URL: https://my.domain.tld/my_prefix/remote.php/webdav/
Bookmarks file path: bookmarks.xbel

### Steps to reproduce
1. installed
2. setup
3. tried syncing

### Expected outcome
bookmarks.xbel should have been saved

### Actual outcome
bookmarks.xbel was not saved

Initially I created an empty bookmarks.xbel on my server and the sync would halt due to `Account.js:171 Syncing failed with Cannot read property 'childNodes' of undefined`. This indicates read access is working with my settings, only the file was invalid.

Then I removed it, and I now see log messages for bookmark parsing (COMPARE, CREATE etc)

However, the process fails at PUT/DELETE requests for bookmarks.xbel and bookmarks.xbel.lock with HTTP500.

![](https://i.imgur.com/xaGM1TA.png)

Example log from my owncloud instance:
```
{
    "reqId":"W4EWHcKHVXsAAD7no0AAAAAG",
    "level":3,
    "time":"2018-08-25T08:41:01+00:00",
    "remoteAddr":"<censored>",
    "user":"--",
    "app":"remote",
    "method":"PUT",
    "url":"\/my_prefix\/remote.php\/webdav\/bookmarks.xbel.lock",
    "message":"Exception: {\"Exception\":\"InvalidArgumentException\",
        \"Message\":\"Invalid url "null"\",\"Code\":0,
        \"Trace\":\"#0 /lib/public/Util.php(809):
        OCP\Util::getFullDomain('null')
        #1 /apps/dav/lib/Connector/Sabre/CorsPlugin.php(98): OCP\Util::isSameDomain('null', 'https://my.domain....')
        #2 /lib/composer/sabre/dav/lib/DAV/Server.php(407): OCA\DAV\Connector\Sabre\CorsPlugin->initialize(Object(OCA\DAV\Connector\Sabre\Server))
        #3 /apps/dav/lib/Connector/Sabre/ServerFactory.php(105): Sabre\DAV\Server->addPlugin(Object(OCA\DAV\Connector\Sabre\CorsPlugin))
        #4 /apps/dav/appinfo/v1/webdav.php(57): OCA\DAV\Connector\Sabre\ServerFactory->createServer('/my_prefix/remote...', '/my_prefix/remote...', Object(OCA\DAV\Connector\Sabre\Auth), Object(Closure))
        #5 /remote.php(165): require_once('...')
        #6 {main}\",\"File\":\"/lib/public/Util.php\",\"Line\":775}"
}
```

Looking at this @ the source: https://github.com/owncloud/core/blob/d5836485ac9802db343b6c9eed986a24b50eb072/apps/dav/lib/Connector/Sabre/CorsPlugin.php#L98

It seems owncloud expects the Origin header to be missing or parsable, but `null` isn't.

AFAIK, it is impossible to set with fetch since it's CORS-related :(

Can this be solved on your end or do I need to file this on their repo?

**Comment by jlbprof** [comment]
Hmm, I have never seen that.   I am interested to see if the problem is
Owncloud specific or not.  You have the latest Owncloud as far as I can
tell.  I do not experience the problem with the latest Nextcloud.
 Nextcloud is a fork of Owncloud,  but they both use SabreDAV under the
hood for WebDAV support, so I would expect that they would perform the same.

I just ran across this document:

https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy

It seems related.  I do not see anything that specifies the "Origin"
header, but it does seem like it might be controlled by this directive.   I
do not know if the app sets these directives or it is set in the browser or
what.

Julian

**Comment by minj** [comment]
Re:headers, see https://developer.mozilla.org/en-US/docs/Glossary/Forbidden_header_name
I am not sure, how much of this applies in extensions.

If you take a look at https://github.com/nextcloud/server/tree/master/apps/dav/lib/Connector/Sabre
there is CorsPlugin is nextcloud at all.

I'll try adding `credentials=include` or `mode=no-cors|same-origin` for all fetch requests in floccus as a test. I have no other ideas at this point

**Comment by marcelklehr** [comment]
> mode=no-cors|same-origin for all fetch requests in floccus

I thought the same thing. Still might not work. Could you test in firefox as well to see if this is a bug in chrome?

**Comment by minj** [comment]
I was not able to remove Origin header in chrome.

Firefox is sending `moz-extension://fe404075-445f-4b57-8f4b-0a96fe99639c` as origin

This fails as well with "Only http based URLs supported".

I will file a bug report @ owncloud

**Comment by minj** [comment]
nevermind, it's already fixed @ owncloud/core#32120 :smile:
